### DISCUSSION TITLE
How to efficiently concat 2 previous commands in `ble.sh` ?

### DISCUSSION THREAD

**Discussion by vaab** [akinomyoga/ble.sh#499] [Q&A]
Let's say you type a long command, then a second one. You then realize that you want to chain them. Before ble.sh, I would:
1. start by recalling the first command, add " &&" at the end, press ENTER, 
2. then I would have new prompt with full history access, which would help me easily find my second command. Then I would need only to press ENTER.

I'm new to `ble.sh`, but I can't see a simple way to simply concat this way 2 commands. The multiline editor kicks in step 2. and forbids me apparently to do any history search. What am I doing wrong ? Can I disable multiline editor for this case ? Or can I call the history to insert exactly one entry of the history inside the multiline editor ?

Sorry if I missed something obvious.

**Comment by akinomyoga** [comment]
> How to efficiently concat 2 previous commands in `ble.sh` ?

If you want to connect the two exactly previous commands, you can use the history expansion feature: You can input `!-2 && !! `, and then it is expanded to `<second previous command> && <the previous command> `. Or you can navigate to the second previous command, move the cursor at the end, and insert ` && !! `. Then, the same result can be obtained.

> The multiline editor kicks in step 2. and forbids me apparently to do any history search.

However, it is true that you cannot use the history "search" when you don't know the offset of the command in the command history.

> Can I disable multiline editor for this case ?

I've never thought about the possibility that users would want to disable the feature, so I've never thought about implementing it.

What seems non-trivial is the way to parse such separated commands. I don't want to provide the wrong syntax highlighting because it might confuse the users and cause destructive results unwanted by the user. But then, to provide the correct syntax highlighting, one needs to carry the incomplete syntax tree of the previous command line and somehow start parsing the current command line from an incomplete syntax context. To efficiently implement the syntax analysis in Bash script, ble.sh uses a very specific way of representing the incremental syntax structure, but I didn't design the parser so that it can be used to start the parsing from inside the non-trivial syntax state such as the nested structure. Then, I need to rewrite the Bash parser from scratch.

> Or can I call the history to insert exactly one entry of the history inside the multiline editor ?

That is technically the history expansion. If you know the offset of the entry in the command history, you can write e.g. `!-n` (where *n* is the offset from the latest command line) and press <kbd>Space</kbd> to perform the history expansion. For more details about the history expansion, you can check the reference manual of Bash.

However, as I already wrote, if you do not know the offset of the history entry to insert, you probably want to navigate through the command history. In that case, one way is to copy and paste the command to insert.

1. First you go to the first command in the command history and insert ` && `.
2. Then, you can go to the second command in the comamand history and press <kbd>C-u C-k</kbd> to cut the entire line.
3. Finally, you can come back to the end of the command line of step 1 and paste the line cut in step 2 by pressing <kbd>C-y</kbd>.

If you just want to navigate through the history in-place, you'll need to implement the feature by yourself. Actually, the existing feature `nsearch` performs an essentially similar thing but for the entire command line. We could add an option to `nsearch` so that `nsearch` operates on a part of the current command line (i.e., only the current line in the multiline command line).

**Reply by vaab** [reply]
Thanks for taking the time for that very intricate answer, and letting us understand the constraints underpinning this issue. As you hinted, the workaround you kindly suggested will not cover variations that are pretty common at least in my use case where I just want to build longer commands (`for` and `while` loops, or simply handpick some specific lines from your history to build a complex command, the concatenation of last commands being a very simple example). Knowing that you can fetch in your history each time you type ENTER can be a very potent shortcut.

I'll have to take the time to dig in your suggestion around `nsearch`, as I'm not yet familiar with the overall structure of `ble.sh`.

**Reply by akinomyoga** [reply]
The mentioned `nsearch` switch seems useful even if it is not what you expect, so I think I'll implement it anyway.

For your purpose, maybe a simpler solution is sufficient. One can add a widget that appends the current history entry to the last-edited history entry and go to the last-edited entry.

**Reply by akinomyoga** [reply]
> The mentioned `nsearch` switch seems useful even if it is not what you expect, so I think I'll implement it anyway.

I implemented this in commit 32f290df78554e22d7ff779d34f1dfec271c16df. I added new values of [the `action` option of the `nsearch` widgets](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A74-Editing#user-content-widget-history-search):

> - `action=load-line` ... Replace the current line with the text content of the history entry
> - `action=load-command` ... Replace the current command with the text content of the history entry
> - `action=insert` ... Insert the text content of the history entry in the current position
> - `action=insert-line` ... Insert the text content of the history entry as a new line in the current position

For example, you can rebind <kbd>prior</kbd> (PageUp) and <kbd>next</kbd> (PageDown) to the history search with `action=load-command`:

```console
$ ble-bind -f prior 'history-search-backward action=load-command'
$ ble-bind -f next 'history-search-forward action=load-command'
```

Then, you can press <kbd>prior</kbd><kbd>prior</kbd> to recall the second previous command, input ` && `, and then press <kbd>prior</kbd> to insert the previous command. Schematically, you can do the following:

```console
$ [prior][prior] && [prior][RET]
```

The option `action=load-command` will replace the part of the current command line that is identified as a simple command in the shell syntax. If you instead want to replace the current *line* with the command in the history, you can use `action=load-line`. You may also try other values `action=insert` and `action=insert-line`.

----

> For your purpose, maybe a simpler solution is sufficient. One can add a widget that appends the current history entry to the last-edited history entry and go to the last-edited entry.

I also added the above one in https://github.com/akinomyoga/blesh-contrib/commit/c620f527d7a021ed72e2d1bb9562e8cddcc0a2bb as another solution. This defines a widget `github499/append-to-last-modified`. For example, you can bind it to <kbd>M-insert</kbd> using the following setup:

```bash
# blerc

ble-import config/github499-append-to-last-modified
ble-bind -f 'M-insert' github499/append-to-last-modified
```

With this setting, after navigating through the command history, you can press <kbd>M-insert</kbd>. Then, the current line content will be appended to the last modified history entry. For example, you can do the following to run `echo hello && echo world`:

```console
$ echo hello
hello
$ echo world
world
$ [up][up][M-insert] && [up][M-insert][RET]
```
